aeppl/truncation.py

```diff
@@ -0,0 +1,95 @@
+from typing import List, Optional
+
+import aesara.tensor as at
+import numpy as np
+from aesara.assert_op import Assert
+from aesara.graph.basic import Node
+from aesara.graph.fg import FunctionGraph
+from aesara.graph.opt import local_optimizer
+from aesara.scalar.basic import Clip
+from aesara.tensor.elemwise import Elemwise
+from aesara.tensor.random.op import RandomVariable
+
+from aeppl.logprob import _logcdf, _logprob
+
+
+# TODO: Add interval transform
+class CensoredRV(RandomVariable):
+    r"""A base class for censored `RandomVariable`\s."""
+
+    def __init__(self, *args, base_op, **kwargs):
+        self.base_op = base_op
+        super().__init__(*args, **kwargs)
+
+
+@_logprob.register(CensoredRV)
+def censor_logprob(op, value, *inputs, name=None, **kwargs):
+
+    *rv_params, lower_bound, upper_bound = inputs
+    logprob = _logprob(op.base_op, value, *rv_params, name=name, **kwargs)
+    logcdf = _logcdf(op.base_op, value, *rv_params, name=name, **kwargs)
+
+    # TODO: Check constant -inf
+    # TODO: Exact bound check might be problematic
+    res = at.switch(
+        at.eq(value, lower_bound),
+        logcdf,
+        at.switch(
+            at.eq(value, upper_bound),
+            at.log(1 - at.exp(logcdf)),
+            logprob,
+        ),
+    )
+    res = at.switch(
+        at.or_(at.lt(value, lower_bound), at.gt(value, upper_bound)),
+        -np.inf,
+        res,
+    )
+
+    res = Assert("lower_bound <= upper_bound")(
+        res, at.all(at.le(lower_bound, upper_bound))
+    )
+
+    if name:
+        res.name = f"{name}_censored_logprob"
+
+    return res
+
+
+@local_optimizer(tracks=[Elemwise])
+def censor_rvs(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return
+
+    # TODO: Allow for one sided-truncation with set_subtensor graph (x[x>ub] = ub)
+    if not isinstance(node.op, Elemwise) or not isinstance(node.op.scalar_op, Clip):
+        return
+
+    base_var, lower_bound, upper_bound = node.inputs
+    base_op = base_var.owner.op
+
+    if not isinstance(base_op, RandomVariable):
+        return
+
+    # clipped_var = node.ouputs[0]
```

Comment:
For some weird reason calling `node.outputs` fails, saying `Apply` does not have outputs attribute.

Reply:
Apparently for `Elemwise` nodes it should be `out` and not `outputs`? Also, is there a way to target the `Clip` op directly and not the `Elemwise`?

Reply:
You've sorted this one out, no?

Reply:
> Apparently for `Elemwise` nodes it should be `out` and not `outputs`?

The `outputs` attribute should always be available on an `Apply` node.  If not, there's a serious bug somewhere.

Reply:
> is there a way to target the `Clip` Op directly instead of the `Elemwise`?

If only one instance of the `Elemwise` + `Clip` combination is used (i.e. it's a singleton), then you can target that instance directly; otherwise, no, there's currently no easy (e.g. a single `isinstance`) type-based way to identify this combination.

If we implement `Op` singletons at the metaclass level, this situation could be simplified a lot.  Both alternatively and in addition to that, we could consider overriding `__instancecheck__` so that `isinstance(node.op, at.clip)` will check `issubclass(type(node.op), Elemwise) and isinstance(node.op.scalar_op, Clip)` (or perhaps use strict equality).

Reply:
> > Apparently for `Elemwise` nodes it should be `out` and not `outputs`?
> 
> The `outputs` attribute should always be available on an `Apply` node. If not, there's a serious bug somewhere.

I might have confused and typed `output` singular without realizing it (but type the correct `outputs` during interactive debugging, which confused me). Things seem fine!

Reply:
> If we implement `Op` singletons at the metaclass level, this situation could be simplified a lot. Both alternatively and in addition to that, we could consider overriding `__subclasshook__`/`__subclasscheck__` so that `issubclass(type(node.op), at.clip)` will check `issubclass(type(node.op), Elemwise) and isinstance(type(node.op).scalar_op, Clip)` (or perhaps use strict equality). We would need to make `scalar_op` a class-level attribute, though.

That sounds really useful. In most of the rewrites / opts, I find myself having to do these instance checks manually